text/0119-dynamic-component-creation.md

```diff
@@ -0,0 +1,338 @@
+---
+title: Dynamic Component Creation
+status: DRAFTED
+created_at: 2022-12-21
+updated_at: 2023-01-04
+pr: (leave this empty until the PR is created)
+---
+
+# Dynamic Component Creation
+
+## Summary
+
+This RFC supersedes the existing `lwc:dynamic` directive to introduce a new way of creating dynamic components without reusing the custom element name.
+
+Dynamic components refer to custom elements where the constructor is not known at compile time.
+
+*This proposal only focuses on the dynamic component creation process, lazy loading which was included in the [original proposal for dynamic components](https://rfcs.lwc.dev/rfcs/lwc/0110-dynamic-components) is not included.*
+
+## Basic example
+
+```html
+<template>
+    <div class="container">
+        <lwc:dynamic lwc:is={lazyConstructor}></lwc:dynamic>
```

Comment:
`lwc:is` was defined because at the time we thought the `is` property in web components was going to be implemented everywhere, at this point, we just don't know. The question is: should we use a more descriptive directive? `<lwc:dynamic component:is={}>` or should we stick to it? I think we already agree that all directives should be prefixed with `lwc:` so another option is: `<lwc:dynamic lwc:component={}>` which probably reads a little better.

Reply:
Bikeshedding, but I'm partial to

```html
<lwc:component lwc:this={}>
```

or

```html
<lwc:component lwc:constructor={}>
```

Agreed that `lwc:is` may give the wrong impression. Even though I do like how it reads ("component is ___"):

```html
<lwc:component lwc:is={}>
```

Reply:
I don't have a strong opinion on the name. I like the `<lwc:component lwc:is={}>` combo since it maps to a concept that has been speced out even if all the browsers do not implement it. On the other hand, the `lwc:this` isn't very intuitive in my opinion.

Reply:
In that case, I would vote for `lwc:is` despite the potential confusion. At least it is prefixed by `lwc:`, so it's clearly not the same as the browser one.

Reply:
My vote would go to `<lwc:component lwc:is={}>` as well, I think it conveys the meaning the best out of the other options.

Vue also does something similar with the [`:is`](https://vuejs.org/guide/essentials/component-basics.html#dynamic-components) attribute.

Reply:
`<lwc:component lwc:is={}>` is the winner!